neurolib/models/multimodel/builder/base/network.py

```diff
@@ -0,0 +1,717 @@
+        Gather callbacks from all masses.
+        """
+        return sum([mass._callbacks() for mass in self.masses], [])
+
+    def _numba_callbacks(self):
+        """
+        Gather callbacks from all masses.
+        """
+        return sum([mass._numba_callbacks() for mass in self.masses], [])
+
+    def _sync(self):
+        """
+        Define synchronisation step for the node. This should define the helper
+        and its equations. Must be symbolic, i.e. defined with basic math
+        operators and symengine operations on state vector. Should return list
+        as
+        [(se.Symbol, <symbolic definition>)]
+        """
+        raise NotImplementedError
+
+    @property
+    def initial_state(self):
+        """
+        Return initial state of this node, i.e. sum of masses initial states.
+        """
+        return np.array(sum([mass.initial_state for mass in self], [],))
+
+    def all_couplings(self, mass_indices=None):
+        """
+        Return all couplings from masses denoted by index.
+
+        :param mass_indices: indices as to which mass to probe
+        :type mass_indices: list|None
+        :return: coupling from all masses indicated in the mass_indices
+        :rtype: dict
+        """
+        mass_indices = mass_indices or np.arange(len(self.masses)).tolist()
+        all_couplings = {}
+        var_idx = 0
+        for mass_idx, mass in enumerate(self.masses):
+            if mass_idx in mass_indices:
+                all_couplings.update({var_idx + k: v for k, v in mass.coupling_variables.items()})
+            var_idx += mass.num_state_variables
+        return all_couplings
+
+    def _derivatives(self, network_coupling=None):
+        """
+        Gather all derivatives from all masses.
+
+        :param network_coupling: dict of network coupling for this node, if
+            None, default network coupling will be used
+        :type network_coupling: dict|None
+        :return: derivatives of the state vector for this node
+        :rtype: list
+        """
+        if network_coupling is None:
+            network_coupling = self.default_network_coupling
+        node_system_equations = []
+        var_idx = 0
+        for mass in self:
+            # get coupling variables
+            coupling_vars = {
+                self._strip_index(key): value
+                for key, value in self.sync_symbols.items()
+                if self._strip_index(key) in mass.required_couplings
+            }
+            coupling_vars.update(
+                {key: value for key, value in network_coupling.items() if key in mass.required_couplings}
+            )
+            mass.idx_state_var = self.idx_state_var + var_idx
+            node_system_equations += mass._derivatives(coupling_vars)
+            var_idx += mass.num_state_variables
+        assert len(node_system_equations) == self.num_state_variables
+        return node_system_equations
+
+
+class SingleCouplingExcitatoryInhibitoryNode(Node):
```

Comment:
this is just helper class - the  base `Node` class represents *any* node - this one represent node in which all the masses are coupled via one variable, usually firing rate... this helper then computes the coupling and you do not have to